File: bin/json-tools.js
#!/usr/bin/env node

const fs = require("fs");

function usage() {
  return [
    "Usage: json-tools <validate|beautify|uglify|escape|unescape> [text]",
    "",
    "If text is omitted, input is read from stdin.",
  ].join("\n");
}

function trimJsonText(text) {
  return text.trim().replace(/(?:^[\n\t\r]|[\n\t\r]$)/g, "");
}

function parseValidJson(text) {
  const trimmed = trimJsonText(text);
  try {
    const value = JSON.parse(trimmed);
    return value !== undefined &&
      (value === null || Array.isArray(value) || typeof value === "object")
      ? value
      : undefined;
  } catch {
    return undefined;
  }
}

function isValid(text) {
  return parseValidJson(text) !== undefined;
}

function escapeJson(text) {
  const trimmed = trimJsonText(text);
  return isValid(trimmed)
    ? JSON.stringify(trimmed).replace(/^"/g, "").replace(/"$/g, "")
    : trimmed;
}

function unescapeJson(text) {
  const trimmed = trimJsonText(text);
  let formatted = trimmed;

  try {
    if (!trimmed.startsWith('"')) {
      formatted = `"${formatted}`;
    }
    if (!trimmed.endsWith('"')) {
      formatted = `${formatted}"`;
    }
    return JSON.parse(formatted);
  } catch {
    return trimmed;
  }
}

function beautifyJson(text, indent = 2) {
  const trimmed = trimJsonText(text);
  const value = parseValidJson(trimmed);
  return value !== undefined ? JSON.stringify(value, null, indent) : trimmed;
}

function uglifyJson(text) {
  const trimmed = trimJsonText(text);
  const value = parseValidJson(trimmed);
  return value !== undefined ? JSON.stringify(value) : trimmed;
}

function readInput(args) {
  if (args.length > 0) {
    return args.join(" ");
  }
  return fs.readFileSync(0, "utf8");
}

function main(argv) {
  const [command, ...args] = argv;
  if (!command || command === "-h" || command === "--help") {
    console.log(usage());
    return 0;
  }

  const input = readInput(args);
  switch (command) {
    case "validate":
      console.log(isValid(input) ? "Valid JSON" : "Invalid JSON");
      return isValid(input) ? 0 : 1;
    case "beautify":
      process.stdout.write(beautifyJson(input));
      return 0;
    case "uglify":
      process.stdout.write(uglifyJson(input));
      return 0;
    case "escape":
      process.stdout.write(escapeJson(input));
      return 0;
    case "unescape":
      process.stdout.write(unescapeJson(input));
      return 0;
    default:
      console.error(usage());
      return 2;
  }
}

if (require.main === module) {
  process.exitCode = main(process.argv.slice(2));
}

module.exports = {
  beautifyJson,
  escapeJson,
  isValid,
  uglifyJson,
  unescapeJson,
};

File: src/json_tools.rs
use serde_json::Value;

pub fn trim_json_text(text: &str) -> &str {
    text.trim().trim_matches(['\n', '\t', '\r'])
}

fn is_object_like(value: &Value) -> bool {
    match value {
        Value::Object(_) | Value::Array(_) | Value::Null => true,
        _ => false,
    }
}

pub fn parse_valid_json(text: &str) -> Option<Value> {
    let value = serde_json::from_str::<Value>(trim_json_text(text)).ok()?;
    is_object_like(&value).then_some(value)
}

pub fn validate(text: &str) -> Result<(), String> {
    let value =
        serde_json::from_str::<Value>(trim_json_text(text)).map_err(|error| error.to_string())?;
    if is_object_like(&value) {
        Ok(())
    } else {
        Err("expected a JSON object, array, or null".to_string())
    }
}

pub fn is_valid(text: &str) -> bool {
    parse_valid_json(text).is_some()
}

pub fn escape(text: &str) -> String {
    let trimmed = trim_json_text(text);
    if !is_valid(trimmed) {
        return trimmed.to_string();
    }

    serde_json::to_string(trimmed)
        .unwrap_or_else(|_| trimmed.to_string())
        .trim_start_matches('"')
        .trim_end_matches('"')
        .to_string()
}

pub fn unescape(text: &str) -> String {
    let trimmed = trim_json_text(text);
    let mut quoted = trimmed.to_string();

    if !trimmed.starts_with('"') {
        quoted.insert(0, '"');
    }

    if !trimmed.ends_with('"') {
        quoted.push('"');
    }

    serde_json::from_str::<String>(&quoted).unwrap_or_else(|_| trimmed.to_string())
}

fn reindent_pretty_json(pretty: &str, indent: &str) -> String {
    let mut output = String::with_capacity(pretty.len());
    let mut chars = pretty.chars().peekable();

    while let Some(ch) = chars.next() {
        output.push(ch);
        if ch != '\n' {
            continue;
        }

        let mut depth = 0;
        while chars.next_if_eq(&' ').is_some() {
            depth += 1;
        }

        for _ in 0..(depth / 2) {
            output.push_str(indent);
        }
    }

    output
}

pub fn beautify_with_indent(text: &str, indent: &str) -> String {
    let trimmed = trim_json_text(text);
    let Some(value) = parse_valid_json(trimmed) else {
        return trimmed.to_string();
    };

    let pretty = serde_json::to_string_pretty(&value).unwrap_or_else(|_| trimmed.to_string());
    if indent == "  " {
        return pretty;
    }

    reindent_pretty_json(&pretty, indent)
}

pub fn beautify(text: &str, indent_width: usize) -> String {
    let indent = " ".repeat(indent_width.max(1));
    beautify_with_indent(text, &indent)
}

pub fn uglify(text: &str) -> String {
    let trimmed = trim_json_text(text);
    let Some(value) = parse_valid_json(trimmed) else {
        return trimmed.to_string();
    };

    serde_json::to_string(&value).unwrap_or_else(|_| trimmed.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validates_json_objects_arrays_and_null() {
        assert!(is_valid(r#"{"a":1}"#));
        assert!(is_valid(r#"[1,2]"#));
        assert!(is_valid("null"));
        assert!(!is_valid(r#""text""#));
        assert!(!is_valid("42"));
        assert!(!is_valid("{"));
    }

    #[test]
    fn transforms_json_text() {
        assert_eq!(beautify(r#"{"a":1}"#, 2), "{\n  \"a\": 1\n}");
        assert_eq!(
            beautify(r#"{"a":{"b":1}}"#, 4),
            "{\n    \"a\": {\n        \"b\": 1\n    }\n}"
        );
        assert_eq!(
            beautify_with_indent(r#"{"a":{"b":1}}"#, "\t"),
            "{\n\t\"a\": {\n\t\t\"b\": 1\n\t}\n}"
        );
        assert_eq!(uglify("{\n  \"a\": 1\n}"), r#"{"a":1}"#);
        assert_eq!(escape(r#"{"a":"b"}"#), r#"{\"a\":\"b\"}"#);
        assert_eq!(unescape(r#"{\"a\":\"b\"}"#), r#"{"a":"b"}"#);
    }

    #[test]
    fn preserves_arbitrary_precision_numbers() {
        let json = r#"{"id":9007199254740993123456789}"#;
        assert_eq!(uglify(json), json);
        assert_eq!(
            beautify(json, 2),
            "{\n  \"id\": 9007199254740993123456789\n}"
        );
    }

    #[test]
    fn leaves_invalid_text_unchanged_after_trimming() {
        assert_eq!(beautify("  nope  ", 2), "nope");
        assert_eq!(uglify("  nope  "), "nope");
        assert_eq!(escape("  nope  "), "nope");
        assert_eq!(unescape("  \\q  "), "\\q");
    }
}
